### PR TITLE
Feat/deep swapper

### PR DESCRIPTION
**Pull request by harisreedhar**

### CODE REVIEW

File: facefusion/vision.py [hunk 1]
```diff
@@ -228,6 +234,18 @@ def equalize_frame_color(source_vision_frame : VisionFrame, target_vision_frame
...
 
 
+def calc_histogram_difference(source_vision_frame : VisionFrame, target_vision_frame : VisionFrame) -> float:
+	histogram_source = cv2.calcHist([cv2.cvtColor(source_vision_frame, cv2.COLOR_BGR2HSV)], [ 0, 1 ], None, [ 50, 60 ], [ 0, 180, 0, 256 ])
+	histogram_target = cv2.calcHist([cv2.cvtColor(target_vision_frame, cv2.COLOR_BGR2HSV)], [ 0, 1 ], None, [ 50, 60 ], [ 0, 180, 0, 256 ])
+	histogram_differnce = float(numpy.interp(cv2.compareHist(histogram_source, histogram_target, cv2.HISTCMP_CORREL), [ -1, 1 ], [ 0, 1 ]))
+	return histogram_differnce
+
+
+def blend_vision_frames(source_vision_frame : VisionFrame, target_vision_frame : VisionFrame, factor : float) -> VisionFrame:
```
Comment (henryruhs) [inline facefusion/vision.py:244]: factor is histogram_factor

File: facefusion/vision.py [hunk 2]
```diff
@@ -228,6 +234,18 @@ def equalize_frame_color(source_vision_frame : VisionFrame, target_vision_frame
 	return target_vision_frame
 
 
+def calc_histogram_difference(source_vision_frame : VisionFrame, target_vision_frame : VisionFrame) -> float:
```
Comment (henryruhs) [inline facefusion/vision.py:237]: 1. use this in the unit test for match_frame_color
2. add a stand alone unit testing for this method

File: facefusion/uis/components/deep_swapper_options.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+def listen() -> None:
+	DEEP_SWAPPER_MODEL_DROPDOWN.change(update_face_enhancer_model, inputs = DEEP_SWAPPER_MODEL_DROPDOWN, outputs = DEEP_SWAPPER_MODEL_DROPDOWN)
+
+	processors_checkbox_group = get_ui_component('processors_checkbox_group')
+	if processors_checkbox_group:
+		processors_checkbox_group.change(remote_update, inputs = processors_checkbox_group, outputs = DEEP_SWAPPER_MODEL_DROPDOWN)
+
+
+def remote_update(processors : List[str]) -> gradio.Dropdown:
+	has_face_enhancer = 'deep_swapper' in processors
```
Comment (henryruhs) [inline facefusion/uis/components/deep_swapper_options.py:35]: has face enhancer? ;D

File: facefusion/uis/components/deep_swapper_options.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+	return gradio.Dropdown(visible = has_face_enhancer)
+
+
+def update_face_enhancer_model(face_enhancer_model : FaceEnhancerModel) -> gradio.Dropdown:
```
Comment (henryruhs) [inline facefusion/uis/components/deep_swapper_options.py:39]: again face enhancer here all over the place, wrong names and typings

File: facefusion/processors/modules/deep_swapper.py [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+	model_context = __name__ + '.' + state_manager.get_item('deep_swapper_model')
+	inference_manager.clear_inference_pool(model_context)
+
+
+def get_model_options() -> ModelOptions:
+	deep_swapper_model = state_manager.get_item('deep_swapper_model')
+	return MODEL_SET.get(deep_swapper_model)
+
+
+def get_model_size() -> Size:
```
Comment (henryruhs) [inline facefusion/processors/modules/deep_swapper.py:67]: Remove this again and use size in model set

File: facefusion/vision.py [hunk 6]
```diff
@@ -210,6 +210,12 @@ def normalize_frame_color(vision_frame : VisionFrame) -> VisionFrame:
 	return cv2.cvtColor(vision_frame, cv2.COLOR_BGR2RGB)
 
 
+def adaptive_match_frame_color(source_vision_frame : VisionFrame, target_vision_frame : VisionFrame) -> VisionFrame:
```
Comment (henryruhs) [inline facefusion/vision.py:213]: probably conditional_match_frame_color()

**Review by henryruhs — CHANGES_REQUESTED** [review]

File: facefusion/vision.py (hunk 1 above)
Comment (harisreedhar) [reply facefusion/vision.py:244]: this method doesn't have direct relation with histogram

Comment (harisreedhar) [reply facefusion/vision.py]: i replaced factor with blend_factor

**Pull request merged**